### PR TITLE
Enable detail view of reservations

### LINKED ISSUE #96: [Feature request] Have dedicated landing pages for reservations
As discussed with Mathieu, we have a desire for users to be able to link directly to a page descriptive of certain information within NEMO. Chief among these are reservations, but otherwise usage events and tools could be useful too.

I'm going to take a look at seeing how hard this would be implement, and try to send a PR if I'm successful.

---

We basically want the information in /event_details/reservation/168/, but nicely formatted as an independent page, with links back to the main app, etc.

### PR DESCRIPTION
**Pull request by jat255**
Resolves #96 

---

### Todo:

- [x] modify group table display to use `data_input.items|get_item:key` (requiring `{% load custom_tags_and_filters %}`
  - Still loop the first time to get the key names; rather than second loop, loop through the keys with `get_item` to ensure they're in the same order as the first loop
- [x] Address comments below from @rptmat57  

Some remaining questions:
- modal response from failed callback of changing project doesn't show up until modal is closed (only one modal at a time allowed...)
- the `modal-dialog` class is set in the base template, and so that has to be changed in `templates/base.html` to make all modals `modal-lg` (unless we could make this a parameter somehow)

---

@rptmat57 I'm sure there's a million things I could have done better, but this seems to be working for me. I'm more than happy to take any criticism or modifications. I think I got the title/project change and cancellation working too, although I may have hosed things up there, so I'd be happy for a review.

I'm not sure if there's any testing or anything that needs to be done. I didn't see formal tests defined in the project, but I may have missed something.

Here's what it looks like as staff or the user owning the reservation:
image

And as a different user:
image

### CODE REVIEW

**Comment by jat255** [comment]
FYI, the only changes I made that could interact/interfere with other parts of the code were in [`change_reservation_project`](https://github.com/usnistgov/NEMO/pull/97/files#diff-1759e08770ec21c1991f35533a0710f8f94fa010fabaf04a0223317a04c4d32eR732-R757) of the `calendar.py` view. I noticed that regardless of what happens when you try to change the project, it was returning a 200, so it looked like it succeeded. I just added some condition checks to return various 40X codes if the project could not be changed.

**Comment by rptmat57** [comment]
That was quick!
I really like the look of your reservation details table!

Now for my (very detailed) feedback:

1. We use a limit of 120 characters in .py files NEMO, if you could change that, that'd be great. see https://github.com/usnistgov/NEMO/wiki/Contributions (also run the tests just in case)
2. I would really prefer not to duplicate the reservation details page. Otherwise every time we make a change in the future, we'll have to remember to change it in 2 separate places. That being said, it makes things a bit tricky with the popup and the cancelling etc.

I thought about it a bit, and I think I have an idea on how to make this work. 
A lot of this comes from anticipating similar needs in the future for usage events and other detail views.

I basically want to have a popup template, and be able to use either the full page or the popup template based on a given request parameter. i.e. `http://nemo.nist.gov/event_details/reservation/1/` -> full page, `http://nemo.nist.gov/event_details/reservation/1/?popup_view=true` -> popup

I realize this is slightly outside of the scope of this feature, but if you are open to it, here is what I propose:

1. Change the original reservation details page with your styling (because I like it). and make sure it looks good in the popup.
2. Add a `popup.html` template in `templates/base` folder with the following code:
```
<div class="modal-header">
    <button type="button" class="close" data-dismiss="modal">&times;</button>
    <h4 class="modal-title">{% block title %}{% endblock %}</h4>
</div>
<div class="modal-body">
    {% block content %}{% endblock %}
</div>
```
3. At the top of the reservation_details page, add `{% extends popup_view|yesno:"base/popup.html,base.html" %}`
4. Move the current `reservation_details` function from `calendar.py` into your `details.py` 
5. Rename the file `event_details.py`
6. Move the current reservation details url in `urls.py` to a separate section called `# Event details` right after the `# Calendar` section
7. Change the reservation details function to get a `popup_view` from the request: `popup_view = request.GET.get("popup_view")` and pass it to the template.
8. In `reservation_event_feed.html`, `specific_user_feed.html` and `usage_event_feed.html` change `"{% url 'reservation_details' x.id %}"` to `"{% url 'reservation_details' x.id %}?popup_view=true"`
8. Move the `templates/calendar/reservation_details.html` file to a new folder `templates/event_details/reservation_details.html`
9. Finally, if you want the change title, etc. functions to work, add the js in reservation_details.html with an `{% if not popup_view %}` otherwise change the conditions to only editing if `not popup_view`. I can go either way with having that functionality in there or not.
10. Make sure everything works (I didn't test any of it)

If realize it's a lot, so it's also totally fine if you just do 1) and I'll do the rest.

**Comment by jat255** [comment]
Ok, I believe I've implemented everything you suggested, as well as changing how the reservation question groups are implemented in the template (I did it using a `templatetag`, since it was easier for me to do in Python than in the Django template language). 

There are two remaining issues/things I wanted to point out:

- if viewing details from a modal, and you do something that fails (trying to change project of a past reservation, for example); the modal response from the failed callback doesn't show up until the first modal is closed (since you can only have one modal open at a time...)
- to make the the new table design fit in the modal, I had to make the modal "large" size. This is done by adding `modal-lg` to the `modal-dialog` class, but this is set in the base template, which means all modals will now be `modal-lg` (unless we pulled this out into the `popup.html`, maybe). If this is a problem, we should refactor it a bit

File: NEMO/templatetags/custom_tags_and_filters.py [hunk 1]
```diff
@@ -63,6 +63,25 @@ def navigation_url(url_name, description):
 		return ""
 
 
+@register.simple_tag
+def res_question_tbody(dictionary):
+	input_dict = dictionary[list(dictionary.keys())[0]]
+	headers = list(input_dict.keys())
+	header_cells = ''.join([format_html('<th>{}</th>', h) for h in headers])
+	head_html = format_html('<thead><tr><th>#</th>{}</tr></thead>',
+							mark_safe(header_cells))
```
Comment (rptmat57) [inline NEMO/templatetags/custom_tags_and_filters.py:72]:
```suggestion
	head_html = format_html('<thead><tr><th>#</th>{}</tr></thead>', mark_safe(header_cells))
```

File: NEMO/templatetags/custom_tags_and_filters.py [hunk 2]
```diff
@@ -63,6 +63,25 @@ def navigation_url(url_name, description):
...
+
+	rows = []
+	for i, (index, d) in enumerate(dictionary.items()):
+		data_cells_html = ''.join([format_html("<td>{}</td>", d[h]) for h in headers])
+		row_html = format_html('<tr><th>{}</th>{}</tr>',
+							   i + 1,
+							   mark_safe(data_cells_html))
```
Comment (rptmat57) [inline NEMO/templatetags/custom_tags_and_filters.py:79]:
```suggestion
		row_html = format_html('<tr><th>{}</th>{}</tr>', i + 1, mark_safe(data_cells_html))
```

File: NEMO/templates/event_details/reservation_details.html [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    <table class="table table-striped">
+
+	{% if user.is_staff %}
+    <tr>
+        <td style="vertical-align: middle">Title:</td>
+        <td>
+            <div class="input-group">
+                <input type="text" id="title" class="form-control" placeholder="{{ reservation.user }}" maxlength="200" value="{{ reservation.title }}">
+                <span class="input-group-btn">
+                    <button type="button" class="btn btn-success"
```
Comment (rptmat57) [inline NEMO/templates/event_details/reservation_details.html:71]:
```suggestion
                    <button type="button" class="btn btn-success" data-dismiss="modal"
```

File: NEMO/templates/event_details/reservation_details.html [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                <td style="vertical-align: middle;">Project:</td>
+                <td>
+                    <div class="input-group">
+                        <select id="change_project" name="project" class="form-control">
+                            {% for p in reservation.user.active_projects %}
+                                <option value="{{ p.id }}" {% if reservation.project.id == p.id %}selected{% endif %}>{{ p }}</option>
+                            {% endfor %}
+                        </select>
+                        <span class="input-group-btn">
+                            <button type="button" class="btn btn-success" onclick="{% if popup_view %}$('#dialog').one('hidden.bs.modal', function() { {% endif %} change_reservation_project('{% url 'change_reservation_project' reservation.id %}', $('#change_project').val()) {% if popup_view %} }) {% endif %}">Change project</button>
```
Comment (rptmat57) [inline NEMO/templates/event_details/reservation_details.html:126]:
```suggestion
                            <button type="button" class="btn btn-success" data-dismiss="modal" onclick="{% if popup_view %}$('#dialog').one('hidden.bs.modal', function() { {% endif %} change_reservation_project('{% url 'change_reservation_project' reservation.id %}', $('#change_project').val()) {% if popup_view %} }) {% endif %}">Change project</button>
```

**Comment by rptmat57** [comment]
adding the "data-dismiss" attribute on the buttons takes care of the issue with the new popup not appearing.

for the large popup, I'll take care of it. I would prefer it only be large for this, since all error messages etc. look a bit silly in the big version

File: NEMO/templates/event_details/reservation_details.html [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        <td>{{ reservation.end }}</td>
+    </tr>
+    <tr>
+        <td>Identifier:</td>
+        <td>{{ reservation.id }}</td>
+    </tr>
+    {% if reservation.question_data_json.items %}
+    <tr>
+        <td>Reservation questions:</td>
+        <td>
```
Comment (rptmat57) [inline NEMO/templates/event_details/reservation_details.html:148]:
```suggestion
        <td style="padding: 0;">
```

File: NEMO/templates/event_details/reservation_details.html [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+            <table class="table table-striped">
```
Comment (rptmat57) [inline NEMO/templates/event_details/reservation_details.html:149]:
```suggestion
            <table class="table table-striped" style="margin-bottom: 0;">
```

File: NEMO/templates/event_details/reservation_details.html [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+            {% for question_name, data in reservation.question_data_json.items %}
+                {% if data.type == 'group' %}
+                    <tr>
+                        <td>{{ data.title }}{% if data.title|slice:"-1:" != ":" %}:{% endif %}</td>
+                        <td>
+                            <table class="table table-striped">
```
Comment (rptmat57) [inline NEMO/templates/event_details/reservation_details.html:155]:
```suggestion
                            <table class="table table-striped" style="margin-bottom: 0;">
```

File: NEMO/templates/event_details/reservation_details.html [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    </tr>
+    {% if reservation.question_data_json.items %}
+    <tr>
+        <td>Reservation questions:</td>
+        <td>
+            <table class="table table-striped">
+            {% for question_name, data in reservation.question_data_json.items %}
+                {% if data.type == 'group' %}
+                    <tr>
+                        <td>{{ data.title }}{% if data.title|slice:"-1:" != ":" %}:{% endif %}</td>
```
Comment (rptmat57) [inline NEMO/templates/event_details/reservation_details.html:153]:
```suggestion
                        <td {% if forloop.first %}style="border-top:none;"{% endif %}>{{ data.title }}{% if data.title|slice:"-1:" != ":" %}:{% endif %}</td>
```

File: NEMO/templates/event_details/reservation_details.html [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+                        <td>
```
Comment (rptmat57) [inline NEMO/templates/event_details/reservation_details.html:154]:
```suggestion
                        <td style="padding: 0;{% if forloop.first %}border-top:none;{% endif %}">
```

File: NEMO/templates/event_details/reservation_details.html [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+                            <table class="table table-striped">
+                                {% res_question_tbody data.user_input %}
+                            </table>
+                        </td>
+                    </tr>
+                {% else %}
+                    <tr>
+                        <td>{{ data.title }}{% if data.title|slice:"-1:" != ":" %}:{% endif %}</td>
```
Comment (rptmat57) [inline NEMO/templates/event_details/reservation_details.html:162]:
```suggestion
                        <td {% if forloop.first %}style="border-top:none;"{% endif %}>{{ data.title }}{% if data.title|slice:"-1:" != ":" %}:{% endif %}</td>
```

File: NEMO/templates/event_details/reservation_details.html [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+                        <td>{{ data.user_input|linebreaksbr }}</td>
```
Comment (rptmat57) [inline NEMO/templates/event_details/reservation_details.html:163]:
```suggestion
                        <td {% if forloop.first %}style="border-top:none;"{% endif %}>{{ data.user_input|linebreaksbr }}</td>
```

**Comment by jat255** [comment]
Ok, cool. Let me know if you need any changes from me. You should be able to push to my branch directly, I think

**Review by rptmat57 — CHANGES_REQUESTED** [review]
alright if you can address those things that should do it.
I'll fix the popup size separately
thanks! sorry this got a bit crazy

**Comment by jat255** [comment]
Oh, I see your comments now. I'll take care of those shortly

**Comment by rptmat57** [comment]
most of them can be pushed directly from the suggestions.
there is a couple more that you'll need to address

**Review by jat255 — COMMENTED** [review]
These were addressed in a separate comment (I didn't feel comfortable with knowing how to do it directly from Github, so sorry if that messed up the code-review process)

**Review by rptmat57 — APPROVED** [review]
all good, thanks for being accommodating!!

**Pull request merged**